java-manta-client-unshaded/src/main/java/com/joyent/manta/client/MantaBucketListingIterator.java

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright (c) 2019, Joyent, Inc. All rights reserved.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.joyent.manta.client;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.joyent.manta.exception.MantaIOException;
+import com.joyent.manta.exception.MantaResourceCloseException;
+import com.joyent.manta.http.HttpHelper;
+import com.joyent.manta.util.MantaUtils;
+import org.apache.commons.lang3.Validate;
+import org.apache.http.HttpEntity;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.io.UncheckedIOException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.joyent.manta.util.MantaUtils.formatPath;
+
+/**
+ * <p>Class that wraps the paging of buckets listing in Manta to a single
+ * continuous iterator.</p>
+ *
+ * <p><a href="https://apidocs.joyent.com/manta/api.html#ListBuckets">
+ * Listing Manta Buckets</a> is done by getting pages of data with N number
+ * of records per page. Each page of data is requested using a limit (number
+ * of records) and a marker (the last seen item in the list). This class
+ * automates that process and abstracts out the details of the paging process.</p>
+ *
+ * @author <a href="https://github.com/nairashwin952013">Ashwin A Nair</a>
+ */
+@SuppressWarnings("Duplicates")
+public class MantaBucketListingIterator implements Iterator<Map<String, Object>>,
```

Comment:
Why is the type `Iterator<Map<String, Object>>`? You will always have the same object type as a value within the `Map`.

Reply:
Won't the same `MantaBucketListingIterator` be used to list buckets in the `/usr/buckets/` directory and bucket objects in the user/buckets/$bucket_name/objects/` where both the have different content-type response headers when we do a `GET` API call for listing, which potentially will be `application/json; type=bucket` and `application/json; type=bucketobject` respectively